File: 060_permutation_sequence/Solution2/Solution.py
class Solution:
    def __init__(self):
        self.facts = [1, 1, 2, 6, 24, 120, 720, 5040, 40320, 362880]

    def get_fact(self, n: int) -> int:
        if n < 10: return self.facts[n]
        return n * self.get_fact(n - 1)

    def getPermutation(self, n: int, k: int) -> str:
        result = []
        is_visited = [False] * n
        k -= 1
        for i in range(n):
            tmp_fact = self.get_fact(n - 1 - i)
            cnt = k // tmp_fact
            k %= tmp_fact
            for j in range(n):
                if is_visited[j]: continue
                if cnt == 0:
                    is_visited[j] = True
                    result.append(str(j + 1))
                    break
                cnt -= 1

        return ''.join(result)


if __name__ == '__main__':
    print(Solution().getPermutation(4, 9))

File: 060_permutation_sequence/Solution2/Solution.java
class Solution {
    final int[] facts = {1, 1, 2, 6, 24, 120, 720, 5040, 40320, 362880};

    public int getFactorial(int n) {
        if (n < 10) return facts[n];
        return n * getFactorial(n - 1);
    }

    public String getPermutation(int n, int k) {
        StringBuilder result = new StringBuilder();
        boolean[] isVisited = new boolean[n];
        k--;
        for (int i = 0; i < n; i++) {
            int tmpFact = getFactorial(n - 1 - i);
            int cnt = k / tmpFact;
            k %= tmpFact;
            for (int j = 0; j < n; j++) {
                if (isVisited[j]) continue;
                if (cnt-- == 0) {
                    isVisited[j] = true;
                    result.append(j + 1);
                    break;
                }
            }
        }

        return String.valueOf(result);
    }

    public static void main(String[] args) {
        System.out.println(new Solution().getPermutation(4, 9));
    }
}
